dali/operators/reader/loader/nemo_asr_loader.cc

```diff
@@ -0,0 +1,212 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <string>
+#include "dali/core/common.h"
+#include "dali/core/error_handling.h"
+#include "dali/operators/decoder/audio/generic_decoder.h"
+#include "dali/operators/decoder/audio/audio_decoder_impl.h"
+#include "dali/operators/reader/loader/nemo_asr_loader.h"
+#include "dali/operators/reader/loader/utils.h"
+#include "dali/pipeline/data/views.h"
+#include "dali/core/tensor_shape_print.h"
+#include "dali/core/static_switch.h"
+#include "dali/pipeline/util/lookahead_parser.h"
+#include "dali/util/file.h"
+
+namespace dali {
+
+namespace detail {
+
+std::string trim(const std::string& str,
+                 const std::string& whitespace = " \t") {
+  const auto str_begin = str.find_first_not_of(whitespace);
+  if (str_begin == std::string::npos)
+    return {};  // no content
+  const auto str_end = str.find_last_not_of(whitespace);
+  const auto str_len = str_end - str_begin + 1;
+  return str.substr(str_begin, str_len);
+}
+
+std::string NormalizeText(std::string& text) {
+  // Remove trailing and leading whitespace
+  auto norm_text = trim(text);
+
+  // Convert to lowercase
+  for (auto &c : norm_text) {
+    c = std::tolower(c);
+  }
+  return norm_text;
+}
+
+void ParseManifest(std::vector<NemoAsrEntry> &entries, std::istream& manifest_file,
+                   float min_duration, float max_duration, bool normalize_text) {
+  std::string line;
+  while (std::getline(manifest_file, line)) {
+    detail::LookaheadParser parser(const_cast<char*>(line.c_str()));
+    if (parser.PeekType() != kObjectType) {
+      DALI_WARN(make_string("Skipping invalid manifest line: ", line));
+      continue;
+    }
+    parser.EnterObject();
+    NemoAsrEntry entry;
+    while (const char* key = parser.NextObjectKey()) {
+      if (0 == detail::safe_strcmp(key, "audio_filepath")) {
```

Comment:
`safe_strcmp` may be _safe_, but it's not _correct_ - what it does is actually `startswith`.

Reply:
You are totally right. I'll fix